### PR TITLE
Smowton/merge/master 20170614

### PR DESCRIPTION
**Pull request by smowton**
This merges latest master, using a slightly unusual method:

1. Merge https://github.com/smowton/cbmc/commit/aed3c25fa3981b4c0c98507beccbfb2c8edcbc5d into current test-gen-support (causes a small number of conflicts, mainly around string refinement, all resolved by keeping HEAD (test-gen-support's version) except for goto-programs/builtin_functions.cpp, whch needs a real merge.

2. Merge the single commit which introduces Doxygen: https://github.com/smowton/cbmc/commit/e9349f8793cba8d387806adb475d74d8e152b4c9 (could have used cherry-pick to the same effect). However instead of trying to resolve the merge, simply
```
git checkout aa183422e46d36a84f6ad314bc7804dcfe31fa3b . # Use the SHA of the previous merge
python scripts/do_doc_convert.py
git rm src/solvers/refinement/refined_string_type.cpp # Deleted here, but not yet in master
git commit -a # Submit the result of doxygen'ing TGS at the result of the merge
```

3. Merge the rest of master as usual: `git merge origin/master`

### CODE REVIEW

**Comment by smowton** [comment]
Gah, I see this has already been overtaken by fresh merges. Re-doing this now.

**Comment by smowton** [comment]
Re-done. @thk123 @reuk quick reviews appreciated so this cycle does not repeat!

**Comment by reuk** [comment]
Just looked at this with @thk123 - all looks good apart from a single function block in `goto-programs/interpreter.cpp` line 1003 which should be deleted.

File: src/java_bytecode/java_bytecode_convert_class.h [hunk 1]
```diff
@@ -14,6 +17,7 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "java_bytecode_parse_tree.h"
 #include "java_bytecode_language.h"
+#include "character_refine_preprocess.h"
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.h]: I'm a bit confused by this addition - it appears to stem from this addition: https://github.com/diffblue/cbmc/commit/924e01bd5693f561584ddf67b3ddc77838d698ac but other parts of that commit have not made it across @romainbrenguier could you comment

File: src/java_bytecode/java_bytecode_convert_method.h [hunk 2]
```diff
@@ -6,6 +6,9 @@ Author: Daniel Kroening, kroening@kroening.com
 
 \*******************************************************************/
 
+/// \file
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_method.h]: This file appears to be missing [this commit](https://github.com/diffblue/cbmc/commit/924e01bd5693f561584ddf67b3ddc77838d698ac), is this intentional/desired @romainbrenguier

Comment (smowton) [inline src/java_bytecode/java_bytecode_convert_method.h]: Romain can confirm but this was replaced with the string_refinementt parameters AFAIK

**Comment by smowton** [comment]
Re-done again, waiting for an ok from @thk123 and Travis

**Pull request merged**